tortoisebeacon/weakcoin/weak_coin.go

```diff
@@ -4,290 +4,325 @@ import (
+	epoch            types.EpochID
+	round            types.RoundID
+	smallestProposal []byte
+	allowances       UnitAllowances
+	// nextRoundBuffer is used to optimistically buffer messages from the next round.
+	nextRoundBuffer []Message
+	coins           map[types.RoundID]bool
+}
+
+// Get the result of the coinflip in this round. It is only valid inbetween StartEpoch/EndEpoch
+// and only after CompleteRound was called.
+func (wc *WeakCoin) Get(epoch types.EpochID, round types.RoundID) bool {
 	if epoch.IsGenesis() {
 		return false
 	}
 
-	pair := epochRoundPair{
-		EpochID: epoch,
-		Round:   round,
+	wc.mu.RLock()
+	defer wc.mu.RUnlock()
+	if wc.epoch != epoch {
+		wc.logger.Panic("requested epoch wasn't started or already completed",
+			log.Uint32("started_epoch", uint32(wc.epoch)),
+			log.Uint32("requested_epoch", uint32(epoch)))
 	}
 
-	wc.weakCoinsMu.RLock()
-	defer wc.weakCoinsMu.RUnlock()
-
-	return wc.weakCoins[pair]
+	return wc.coins[round]
 }
 
-func (wc *weakCoin) PublishProposal(ctx context.Context, epoch types.EpochID, round types.RoundID) error {
-	wc.Log.With().Info("Publishing proposal",
-		log.Uint64("epoch_id", uint64(epoch)),
-		log.Uint64("round_id", uint64(round)))
-
-	proposal, err := wc.generateProposal(epoch, round)
-	if err != nil {
-		return err
-	}
-
-	signedProposal := wc.signProposal(proposal)
+// StartEpoch notifies that epoch is started and we can accept messages for this epoch.
+func (wc *WeakCoin) StartEpoch(epoch types.EpochID, allowances UnitAllowances) {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	wc.epoch = epoch
+	wc.allowances = allowances
+}
 
-	if wc.proposalExceedsThreshold(proposal) {
-		// If a proposal exceeds the threshold, it is not sent.
-		return nil
-	}
+// FinishEpoch completes epoch. After it is called Get for this epoch will panic.
+func (wc *WeakCoin) FinishEpoch() {
+	wc.mu.Lock()
+	defer wc.mu.Unlock()
+	wc.allowances = nil
+	wc.coins = map[types.RoundID]bool{}
+	wc.round = 0
+}
 
-	message := Message{
-		Epoch:        epoch,
-		Round:        round,
-		MinerID:      wc.minerID,
-		VRFSignature: signedProposal,
+// StartRound process any buffered messages for this round and broadcast our proposal.
+func (wc *WeakCoin) StartRound(ctx context.Context, round types.RoundID) error {
+	wc.mu.Lock()
+	wc.logger.With().Info("started round",
+		log.Uint32("epoch_id", uint32(wc.epoch)),
+		log.Uint64("round_id", uint64(round)))
+	wc.round = round
+	wc.smallestProposal = nil
+	epoch := wc.epoch
+	for i, msg := range wc.nextRoundBuffer {
+		if msg.Epoch != wc.epoch || msg.Round != wc.round {
+			continue
+		}
+		if err := wc.updateProposal(msg); err != nil {
+			wc.logger.With().Debug("received invalid proposal",
+				log.Err(err),
+				log.Uint32("epoch_id", uint32(msg.Epoch)),
+				log.Uint64("round_id", uint64(msg.Round)))
+		}
+		wc.nextRoundBuffer[i].Epoch = 0
+		wc.nextRoundBuffer[i].Round = 0
+		wc.nextRoundBuffer[i].Unit = 0
+		wc.nextRoundBuffer[i].Signature = nil
```

Comment:
Why did you prefer this over the code below?
```suggestion
		wc.nextRoundBuffer[i] = Message{}
```

Reply:
do you know if this will allocate a new memory or overwrite existing? doesn't matter much to me, but this is what i was thinking when made that change

Reply:
It won't allocate anything because of no pointers. It will do `runtime.typedmemclr`.